source/opt/dominator_analysis.cpp

```diff
@@ -37,5 +39,35 @@ ir::BasicBlock* DominatorAnalysisBase::CommonDominator(
   return block;
 }
 
+bool DominatorAnalysisBase::Dominates(ir::Instruction* a,
+                                      ir::Instruction* b) const {
+  if (!a || !b) {
+    return false;
+  }
+
+  if (a == b) {
+    return true;
+  }
+
+  ir::BasicBlock* bb_a = a->context()->get_instr_block(a);
+  ir::BasicBlock* bb_b = b->context()->get_instr_block(b);
+
+  if (bb_a != bb_b) {
+    return tree_.Dominates(bb_a, bb_b);
+  }
+
+  for (ir::Instruction& inst : *bb_a) {
+    if (&inst == a) {
+      return true;
+    } else if (&inst == b) {
+      return false;
+    }
+  }
```

Comment:
If only one of the instructions has the wrong basic block information, this will silently return true/false.  Perhaps, traverse the whole block always?

Reply:
In general, I do not want to do extra work.  I want to assert when I hit an know problem.  If you really want, I can add a loop that is guarded by an ifdef that can do this work.

Reply:
Actually, I think it's fine as is.  This code is guarded with the predicate bb_a != bb_b, so if get_instr_block() is lying, we have more problems that will lead to other problems elsewhere.